### ISSUE TITLE
Hardware advice for best performance?

### ISSUE THREAD

**Issue by mcqueary** [nats-io/nats-site#187]
_From @monopollly on November 22, 2016 14:35_

Could you please write a hardware preferences for nats...

something like this:
https://www.elastic.co/guide/en/elasticsearch/guide/current/hardware.html

Thanks!

_Copied from original issue: nats-io/gnatsd#380_

**Comment by mcqueary** [comment]
Thanks for the great suggestion! I'm moving this over to the nats-io/nats-site issues list.

**Comment by davisford** [comment]
Any ballpark recommendations for EC2 instance sizes that work well with natsd?  I get that it is application dependent, but I'm just looking for helpful / anecdotal information about people that are using it in production.

**Comment by ColinSullivan1** [comment]
@davisford , this is a very difficult question to answer...  The short answer is that we'd need more information - what requirements you have with respect to throughput and # of subscribers.  Even at that, I'd still suggest you test to see what works for you.

In research for NATS Cloud, we tested - thoroughly - to determine what the minimum size would be to accommodate the load we needed.  This included memory (# of subscribers and connections), and throughput (which is CPU intensive), as well as network requirements.  For example, supporting a dozen servers with a minimum of 200k 512 byte messages/sec, per server, and hundreds of subscribers, might require a m4.10x large, especially if TLS is involved.  Alternatively, if you require only one server for a few subscribers at 1000 msgs/sec, a t2.small (or maybe even a t2.micro!) may work for you.  The NATS server is very lightweight.

The other issue to consider is network; if you are sending and receiving a large amount of data, you could be limited to the instance types that support enhanced networking.

Basically, you'll want to test your limits per your requirements.  I have a rough tool that simulates clients, found [here](https://github.com/ColinSullivan1/nats-testing/tree/master/nats-client-sim), which may provide a starting point for you to test and hone in on the right machine size.

That being said, it'd be great to get additional community input here!  I'm certainly interested in what you find in your case.

Thanks,
Colin

**Comment by davisford** [comment]
Thanks @ColinSullivan1 -- that's helpful.  I did just spawn a 3-node cluster in ECS.  I used t2.medium, and I'm monitoring it to see how it does.

**Comment by duycuong87vn** [comment]
hi, @davisford , @ColinSullivan1 
I need to calculate resources for the system, and I found this issue.
---
On my case.
I setup Nats-Cluster (6 nodes) for 200.000 CCU,
My system needs to serve 200,000 subscribers  (200,000 CCU and keep client connections alive for some minutes... ~ 40 minutes), to listen/receive  1 byte message from publisher.
I'm planning to use m5.large vcpu 2 ram 8gb.  x 6 nodes
Please give your opinion.
Thanks!
